### PR TITLE
fix pod volume passing and alter infra inheritance

### LINKED ISSUE #13548: Podman 4.x Pod volume mounts
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind bug

**Description**
When trying adding containers to a pod that has volume mounts, the containers do not have any of the volumes mounted

**Steps to reproduce the issue:**

1. create pod with a volume mount
```
podman pod create --name=test -v /home/user/test:/test
```

2. then run a container
```podman run --pod=test fedora touch /test/test.file```

**Describe the results you received:**
```touch: cannot touch '/test/test.file': No such file or directory```
The container will say the folder/file does not exist

**Describe the results you expected:**
The container will have the volume mounts from the pod it is a part of

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `podman version`:**
```
podman version 4.0.2
```

**Output of `podman info --debug`:**
```
host:
  arch: amd64
  buildahVersion: 1.24.1
  cgroupControllers:
  - memory
  - pids
  cgroupManager: systemd
  cgroupVersion: v2
  conmon:
    package: conmon-2.1.0-2.fc35.x86_64
    path: /usr/bin/conmon
    version: 'conmon version 2.1.0, commit: '
  cpus: 12
  distribution:
    distribution: fedora
    variant: server
    version: "35"
  eventLogger: journald
  hostname: kuroda.home.pf
  idMappings:
    gidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65536
    uidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65536
  kernel: 5.16.14-200.fc35.x86_64
  linkmode: dynamic
  logDriver: journald
  memFree: 30800465920
  memTotal: 32983187456
  networkBackend: netavark
  ociRuntime:
    name: crun
    package: crun-1.4.3-1.fc35.x86_64
    path: /usr/bin/crun
    version: |-
      crun version 1.4.3
      commit: 61c9600d1335127eba65632731e2d72bc3f0b9e8
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +CRIU +YAJL
  os: linux
  remoteSocket:
    exists: true
    path: /run/user/1000/podman/podman.sock
  security:
    apparmorEnabled: false
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: true
    seccompEnabled: true
    seccompProfilePath: /usr/share/containers/seccomp.json
    selinuxEnabled: false
  serviceIsRemote: false
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: slirp4netns-1.1.12-2.fc35.x86_64
    version: |-
      slirp4netns version 1.1.12
      commit: 7a104a101aa3278a2152351a082a6df71f57c9a3
      libslirp: 4.6.1
      SLIRP_CONFIG_VERSION_MAX: 3
      libseccomp: 2.5.3
  swapFree: 25769795584
  swapTotal: 25769795584
  uptime: 19m 47.53s
plugins:
  log:
  - k8s-file
  - none
  - passthrough
  - journald
  network:
  - bridge
  - macvlan
  volume:
  - local
registries:
  search:
  - registry.fedoraproject.org
  - registry.access.redhat.com
  - docker.io
  - quay.io
store:
  configFile: /home/kuroda/.config/containers/storage.conf
  containerStore:
    number: 2
    paused: 0
    running: 2
    stopped: 0
  graphDriverName: overlay
  graphOptions: {}
  graphRoot: /home/kuroda/.local/share/containers/storage
  graphStatus:
    Backing Filesystem: xfs
    Native Overlay Diff: "true"
    Supports d_type: "true"
    Using metacopy: "false"
  imageCopyTmpDir: /var/tmp
  imageStore:
    number: 4
  runRoot: /run/user/1000/containers
  volumePath: /home/kuroda/.local/share/containers/storage/volumes
version:
  APIVersion: 4.0.2
  Built: 1646943965
  BuiltTime: Thu Mar 10 15:26:05 2022
  GitCommit: ""
  GoVersion: go1.16.14
  OsArch: linux/amd64
  Version: 4.0.2
```
**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**
podman-4.0.2-5.fc35.x86_64

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide? (https://github.com/containers/podman/blob/main/troubleshooting.md)**
Yes

**Additional environment details (AWS, VirtualBox, physical, etc.):**
This is running on my current desktop. The mounts do work when I mount them directly too the container. The latest documentation about the ```podman pod create --volume``` command [here](https://docs.podman.io/en/latest/markdown/podman-pod-create.1.html#volume-v-source-volume-host-dir-container-dir-options) seems to imply that containers in that pod will all share the mounts defined in the create command, but that isn't working in my case. I have tried running with all of the propagation option as well.

### PR DESCRIPTION
**Pull request by cdoern**
the infra Inherit function was not properly passing pod volume information to new containers. alter the inherit function and struct to use the new `ConfigToSpec` function used in clone pick and choose the proper entities from a temp spec and validate them on the spegen side rather than passing directly to a config

resolves #13548

Signed-off-by: cdoern <cbdoer23@g.holycross.edu>

### CODE REVIEW

**Comment by cdoern** [comment]
@containers/podman-maintainers PTAL. This is a pretty substantial bug fix as pod volumes seem to have been broken for some time.

File: pkg/specgen/generate/container.go [hunk 1]
```diff
@@ -483,5 +488,17 @@ func ConfigToSpec(rt *libpod.Runtime, specg *specgen.SpecGenerator, contaierID s
...
 	specg.Mounts = conf.Spec.Mounts
 	specg.HostDeviceList = conf.DeviceHostSrc
+	mapSecurityConfig(conf, specg)
 	return c, nil
 }
+
+// mapSecurityConfig takes a libpod.ContainerSecurityConfig and converts it to a specgen.ContinerSecurityConfig
+func mapSecurityConfig(c *libpod.ContainerConfig, s *specgen.SpecGenerator) {
+	s.Privileged = c.Privileged
+	if len(c.LabelOpts) > 0 {
```
Comment (rhatdan) [inline pkg/specgen/generate/container.go:498]: Is this check necessary,  If you append an empty list, does it fail? Blow up?

**Comment by mheon** [comment]
Cause of errors seems to be https://github.com/containers/podman/blob/main/test/apiv2/40-pods.at#L88 which is pulling from docker.io

@edsantiago @cevich PTAL - is there anything in the APIv2 tests preventing pulls from docker.io? Because that test has been there for 2 years, seemingly, and I'm amazed it only now started blowing up with rate-limiting errors

**Comment by edsantiago** [comment]
@mheon I think we have some magic doohickey that masquerades docker.io, using quay.io instead. I can't remember the details. However, instead of looking them up, I do note that the error I see is:
```
 >Error: initializing source docker://busybox:latest: Requesting bearer token: invalid status code from registry 500 (Internal Server Error)
```
...which is a 500, not the usual "rate limit blah blah". And, looking at the [full log](https://storage.googleapis.com/cirrus-ci-6707778565701632-fcae48/artifacts/containers/podman/4532610264203264/html/apiv2-podman-fedora-35-root-host.log.html#t--00272), I see another 500 error, this time very clearly quay.io. This looks like a quay flake to me.

(No point in pressing Re-run, though, because of so many other errors in this CI run)

File: libpod/container_config.go [hunk 2]
```diff
@@ -406,12 +407,17 @@ type ContainerMiscConfig struct {
...
+	OverlayVolumes     []*specgen.OverlayVolume `json:"overlay_volumes,omitempty"`
+	ImageVolumes       []*specgen.ImageVolume   `json:"image_volumes,omitempty"`
+	InfraResources     *spec.LinuxResources     `json:"resource_limits,omitempty"`
+	CapAdd             []string                 `json:"cap_add,omitempty"`
+	CapDrop            []string                 `json:"cap_drop,omitempty"`
+	SelinuxOpts        []string                 `json:"selinux_opts,omitempty"`
+	ApparmorProfile    string                   `json:"apparmor_profile,omitempty"`
+	SeccompPolicy      string                   `json:"seccomp_policy,omitempty"`
+	SeccompProfilePath string                   `json:"seccomp_profile_path,omitempty"`
+	NoNewPrivileges    bool                     `json:"no_new_privileges,omitempty"`
```
Comment (TomSweeneyRedHat) [inline libpod/container_config.go:422]: Could we put these in alpha order please?

File: pkg/specgen/specgen.go [hunk 3]
```diff
@@ -523,8 +523,8 @@ type ContainerHealthCheckConfig struct {
 // swagger:model SpecGenerator
 type SpecGenerator struct {
 	ContainerBasicConfig
-	ContainerStorageConfig
-	ContainerSecurityConfig
+	ContainerStorageConfig  `json:"container_storage_config,omitempty"`
+	ContainerSecurityConfig `json:"container_security_config,omitempty"`
```
Comment (Luap99) [inline pkg/specgen/specgen.go:527]: Why the change here? I think this will break backwards compat. Spegcen must be stable.

File: pkg/specgen/generate/container_create.go [hunk 4]
```diff
@@ -535,32 +535,68 @@ func createContainerOptions(ctx context.Context, rt *libpod.Runtime, s *specgen.
...
+	}
+
+	err = json.Unmarshal(compatByte, &s.ContainerSecurityConfig)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	err = json.Unmarshal(compatByte, &s.ContainerStorageConfig)
+	if err != nil {
+		return nil, nil, nil, err
```
Comment (Luap99) [inline pkg/specgen/generate/container_create.go:594]: This looks absolutely insane, why do we need that many marshal/unmarshal calls? It is completely unreadable and nobody will understand what this is doing.

Why can't we just assign the options we need instead of using marshal/unmarshal?

Comment (cdoern) [reply pkg/specgen/generate/container_create.go:594]: i will try to find a more succinct form of this. It needs to be unmarshall unless we want a function that is basically just a list of 100 assignments

**Comment by cevich** [comment]
> I think we have some magic doohickey that masquerades docker.io, using quay.io instead

Yes, see `test/registries.conf`

File: pkg/specgen/generate/container_create.go (hunk 4 above)
Comment (cdoern) [reply pkg/specgen/generate/container_create.go]: fixed it up a bit @Luap99, lets see if it works...

File: pkg/specgen/generate/container.go (hunk 1 above)
Comment (cdoern) [reply pkg/specgen/generate/container.go:557]: i will check with and without. I remember it adding a blank label and blowing up.

Comment (cdoern) [reply pkg/specgen/generate/container.go]: it works without the check!

**Comment by cdoern** [comment]
@containers/podman-maintainers  PTAL

File: pkg/specgen/generate/container.go [hunk 5]
```diff
@@ -481,7 +489,68 @@ func ConfigToSpec(rt *libpod.Runtime, specg *specgen.SpecGenerator, contaierID s
 		}
 	}
 	specg.OverlayVolumes = overlay
-	specg.Mounts = conf.Spec.Mounts
+	name, mounts := c.SortUserVolumes(c.Spec())
+	tempInspect, err := c.GetInspectMounts(name, conf.ImageVolumes, mounts)
```
Comment (rhatdan) [inline pkg/specgen/generate/container.go:493]: Hate this function name GetInspectMounts... Nothing to fix though.

**Review by rhatdan — APPROVED** [review]
LGTM other then a Nit/

File: pkg/specgen/generate/container.go (hunk 5 above)
Comment (cdoern) [reply pkg/specgen/generate/container.go:493]: if we are changing `GetMounts` to `Mounts` above I can rename this `GetMounts`

File: pkg/specgenutil/volumes.go [hunk 6]
```diff
@@ -172,7 +172,7 @@ func findMountType(input string) (mountType string, tokens []string, err error)
 // podman run --mount type=bind,src=/etc/resolv.conf,target=/etc/resolv.conf ...
 // podman run --mount type=tmpfs,target=/dev/shm ...
 // podman run --mount type=volume,source=test-volume, ...
-func getMounts(mountFlag []string) (map[string]spec.Mount, map[string]*specgen.NamedVolume, map[string]*specgen.ImageVolume, error) {
+func GetMounts(mountFlag []string) (map[string]spec.Mount, map[string]*specgen.NamedVolume, map[string]*specgen.ImageVolume, error) {
```
[the remark this answers is not in the document]
Comment (cdoern) [reply pkg/specgenutil/volumes.go:175]: Sure, I can change this.

File: pkg/specgen/generate/container.go [hunk 7]
```diff
@@ -481,7 +489,68 @@ func ConfigToSpec(rt *libpod.Runtime, specg *specgen.SpecGenerator, contaierID s
...
 	specg.OverlayVolumes = overlay
-	specg.Mounts = conf.Spec.Mounts
+	name, mounts := c.SortUserVolumes(c.Spec())
+	tempInspect, err := c.GetMounts(name, conf.ImageVolumes, mounts)
+	if err != nil {
+		return nil, nil, err
+	}
+	mountsToPass := []string{}
+	for _, val := range tempInspect {
+		mountsToPass = append(mountsToPass, fmt.Sprintf("type=%v,src=%v,target=%v,%v", val.Type, val.Source, val.Destination, strings.Join(val.Options, ",")))
```
[the remark this answers is not in the document]
Comment (cdoern) [reply pkg/specgen/generate/container.go:499]: so the issue with passing the mounts directly (from `spec.Mounts`) is that you also pick up a large amount of mounts created further down in the process that in specgen cause errors (/proc for example). 

I also think its a good idea to re validate the mounts in case they break something else the user has passed (a new mount for example when I add that option to clone). 

This is basically recreating the `--mount` flag and making sure everything fits correctly. Any other way completely blew up or passed too much info

[the remark this answers is not in the document]
Comment (mheon) [reply pkg/specgen/generate/container.go:499]: I'm not really convinced that a full revalidation of the mount and its options is required; presumably, it was already in Podman and working.

[the remark this answers is not in the document]
Comment (mheon) [reply pkg/specgen/generate/container.go:499]: Why do we need to pull directly from the config? The inspect functions you're using have the details on how to determine what mounts were added by users; a slight alteration would be able to get you the relevant mounts from the config

[the remark this answers is not in the document]
Comment (cdoern) [reply pkg/specgen/generate/container.go:499]: well, the `GetMounts` function returns a "fake" mount list that either way need to be translated to spec.Mounts for the specgen. The issue I was running into is that there were mount options that specgen does *not* like ex: `rbind` which get added later (I think?). That is the main purpose of the loop it is just pulling out invalid options that specgen itself will add later

idk if this is the best way but if you take out the loop we get smashed with a lot of options errors.

[the remark this answers is not in the document]
Comment (mheon) [reply pkg/specgen/generate/container.go:499]: I can definitely see needing to to a pass over the options, but the rest of the volume struct should work without change

[the remark this answers is not in the document]
Comment (cdoern) [reply pkg/specgen/generate/container.go:499]: I was using the errors returned by the `Mounts` function as an indication of what options are causing issues below. Is there any other way to easily pass over these?

[the remark this answers is not in the document]
Comment (mheon) [reply pkg/specgen/generate/container.go:499]: Shouldn't be necessary, you can use the logic that inspect uses (which is relatively simple to duplicate) to determine which mounts to use, and pull them out of the existing spec - as opposed to reusing the existing function entirely

[the remark this answers is not in the document]
Comment (mheon) [reply pkg/specgen/generate/container.go:499]: Eh, maybe I'm being too picky here. @Luap99 @vrothberg Thoughts?

[the remark this answers is not in the document]
Comment (cdoern) [reply pkg/specgen/generate/container.go:499]: whatever everyone prefers I can implement either way same result

[the remark this answers is not in the document]
Comment (cdoern) [reply pkg/specgen/generate/container.go:499]: I am going to have to recreate: 
https://github.com/containers/podman/blob/a416fd6de40820e266e9e1e312c5f13bbb0b189f/pkg/specgenutil/volumes.go#L186-L237

for the options

[the remark this answers is not in the document]
Comment (mheon) [reply pkg/specgen/generate/container.go:499]: My commitment to see this done right is waning. Parsing again kind of offends my sensibilities, but if it works (and based on your tests it definitely does), I'm not really that committed to seeing it changed.

[the remark this answers is not in the document]
Comment (mheon) [reply pkg/specgen/generate/container.go:499]: Basically, if someone else says this looks fine, I'm good to merge

[the remark this answers is not in the document]
Comment (cdoern) [reply pkg/specgen/generate/container.go:499]: ok. @containers/podman-maintainers  PTAL at this discussion. In the mean time I will keep messing around and see if I can implement it your way.

File: pkg/specgen/generate/container.go [hunk 8]
```diff
@@ -481,7 +489,68 @@ func ConfigToSpec(rt *libpod.Runtime, specg *specgen.SpecGenerator, contaierID s
...
+	}
+	m, _, _, err := specgenutil.Mounts(mountsToPass)
+
+	// a lot of options get added later and are hard to translate.
+	// weed those out here
+	for err != nil && strings.Contains(err.Error(), "invalid mount option") {
```
Comment (Luap99) [inline pkg/specgen/generate/container.go:505]: Can we use proper error comparison here with something like errors.Is().
This will break as soon as someone changes the error message.

File: pkg/specgen/generate/container.go [hunk 9]
```diff
@@ -481,7 +489,68 @@ func ConfigToSpec(rt *libpod.Runtime, specg *specgen.SpecGenerator, contaierID s
...
+		errString := err.Error()
+		errString = errString[:strings.LastIndex(errString, ":")]
+		for i, val := range mountsToPass {
+			if strings.Contains(val, errString) {
+				ind := strings.Index(mountsToPass[i], errString) - 1
+				firstH := mountsToPass[i][0:ind]
+				secondH := mountsToPass[i][ind+len(errString)+1:]
+				mountsToPass[i] = firstH + secondH
+			}
+		}
```
Comment (Luap99) [inline pkg/specgen/generate/container.go:515]: I have no idea what this is doing or why such weird manually conversion is needed. Please add at least a comment explaining what it does

File: pkg/specgen/generate/container.go [hunk 10]
```diff
@@ -481,7 +489,68 @@ func ConfigToSpec(rt *libpod.Runtime, specg *specgen.SpecGenerator, contaierID s
 		}
 	}
 	specg.OverlayVolumes = overlay
-	specg.Mounts = conf.Spec.Mounts
+	name, mounts := c.SortUserVolumes(c.Spec())
+	tempInspect, err := c.GetMounts(name, conf.ImageVolumes, mounts)
+	if err != nil {
+		return nil, nil, err
+	}
+	mountsToPass := []string{}
```
Comment (Luap99) [inline pkg/specgen/generate/container.go:497]: preallocate the slice with `make([]string, 0 , len(tempInspect))`

File: pkg/specgen/generate/container.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (Luap99) [reply pkg/specgen/generate/container.go:499]: I agree with @mheon I don't think we should parse again, the error handling below looks dangerous and fragile.

**Comment by cdoern** [comment]
alright, I will go back and redo the mounts tonight

**Comment by cdoern** [comment]
@Luap99 @mheon tested it a bunch and *i think* sort user volumes does enough to prohibit any bad options or volumes from getting through

File: pkg/specgenutil/volumes.go [hunk 11]
```diff
@@ -466,6 +466,7 @@ func getTmpfsMount(args []string) (spec.Mount, error) {
 			// Since Docker ignores this option so shall we.
 			continue
 		default:
+			//fmt.Println()
```
Comment (Luap99) [inline pkg/specgenutil/volumes.go:469]: debug leftover?

File: pkg/specgen/generate/container_create.go [hunk 12]
```diff
@@ -535,32 +535,24 @@ func createContainerOptions(ctx context.Context, rt *libpod.Runtime, s *specgen.
...
+	}
 	options := []libpod.CtrCreateOption{}
-	compatibleOptions := &libpod.InfraInherit{}
 	infraConf := infra.Config()
 	infraSpec := infraConf.Spec
 
-	config, err := json.Marshal(infraConf)
+	var compatByte []byte
+	compatByte, err = json.Marshal(compatibleOptions)
```
Comment (Luap99) [inline pkg/specgen/generate/container_create.go:549]: Any reason to declare compatByte extra?
Why not use `:=` as done before?

**Comment by mheon** [comment]
Rebase should fix the tests

**Comment by cdoern** [comment]
@containers/podman-maintainers PTAL

**Comment by rhatdan** [comment]
LGTM
